sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java

```diff
@@ -0,0 +1,275 @@
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata.functions.params;
+
+import com.google.common.base.Preconditions;
+import io.crate.analyze.symbol.FuncArg;
+import io.crate.types.DataType;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Type handling for function parameters (FuncParams).
+ *
+ * A FuncParam is composed of zero, one, or multiple parameters ({@link Param}s).
+ *
+ * Parameters are linked with each other by passing the same references more than once.
+ *
+ * For example:
+ * {@code
+ *
+ *  Param a = Param.of(ANY);
+ *  Param b = Param.of(INTEGER);
+ *
+ *  FuncParams.of(a, a, b).build();
+ *  // => A function parameters definition with three parameters.
+ *  // The first and the second parameter (a) can be of any type but must have the same type.
+ *  // The third parameter (b) has to be of type integer.
+ * }
+ *
+ * Matching
+ *
+ * Parameters are matched against an argument list by using the match method:
+ * {@code
+ *
+ *  List<Symbol> args = Arrays.asList(Literal.of(1L), Literal.of(1), Literal.of(1))
+ *
+ *  List<DataType> signature = FuncParams.match(args);
+ *  // => The resulting signature does not have to be of the same type as the original symbols
+ *  // but the implementation assures that a valid conversion is derived from the given types.
+ *  // This is done by checking convertibility and type precedence.
+ * }
+ *
+ * Variable Arguments (VarArgs)
+ *
+ * Additional variable arguments can be specified on FuncParams.Builder:
+ * {@code
+ *
+ *      FuncParams.of().withVarArgs(Param.of(INTEGER)).build()
+ *      // => Any number of integers as var args.
+ *      FuncParams.of().withVarArgs(Param.of(STRING), Param.of(INTEGER)).build()
+ *      // => Any number of String followed by an Integer parameter,
+ *      // e.g. "a", 1, "b", 2, ...
+ *
+ *      FuncParams.of().withVarArgs(2, Param.of(INTEGER)).build()
+ *      // => Maximum of two variable integer arguments.
+ *
+ *      FuncParams.of().withIndependentVarArgs(Param.of(ANY)).build()
+ *      // => Unlimited number of arguments which don't have to be of the same type
+ * }
+ *
+ *
+ * This class is thread-safe and can be accessed by multiple threads at once.
```

Comment:
the `parameters` are being mutated, so this is only thread-safe if the `Param` class provides type-safety. I think that should be mentioned (and Param should probably become a final class, to prevent subclassing)

And I'm not sure if this really is thread-safe. E.g if `resetBoundTypes` is called concurrently from multiple threads, this wouldn't work?